File: src/style-library.test.ts
import {describe,it,expect} from 'vitest';
import {STYLE_SORTS,styleMetadata,sortStyles,moveStyle,parseStyleLines,matchStyleImages,styleCoverIndex} from './style-library';
import type {StylePromptPreset} from './types';
const make=(id:string,rating=0,uses=0,date='2026-09-01'):StylePromptPreset=>({id,name:id,prompt:`artist:${id}`,group:'Default',createdAt:date,rating,usageCount:uses});
describe('saved style library',()=>{
 it('normalizes old presets and invalid metadata without changing prompts',()=>{
  expect(styleMetadata({})).toEqual({rating:0,usageCount:0,sortOrder:Number.MAX_SAFE_INTEGER});
  expect(styleMetadata({rating:4.26,usageCount:3.9,sortOrder:-1})).toEqual({rating:4.26,usageCount:3,sortOrder:0});
  expect(styleMetadata({rating:99,usageCount:NaN})).toMatchObject({rating:5,usageCount:0});
 });
 it.each([['default','bac'],['custom','bac'],['rating-desc','abc'],['rating-asc','bca'],['uses-desc','cba'],['uses-asc','abc'],['created-desc','acb'],['created-asc','bca'],['name','abc']])('%s sorts stably', (mode,want)=>{
  const items=[make('b',1,2,'2026-09-01'),make('a',5,0,'2026-09-03'),make('c',1,3,'2026-09-02')];
  expect(sortStyles(items,mode).map(x=>x.id).join('')).toBe(want);expect(items.map(x=>x.id).join('')).toBe('bac');
 });
 it('moves in either direction and preserves original/default order',()=>{
  const original=['a','b','c'].map(x=>make(x));const moved=moveStyle(original,'a','c');
  expect(sortStyles(moved,'custom').map(x=>x.id)).toEqual(['b','c','a']);
  expect(moved.map(x=>x.id)).toEqual(['a','b','c']);expect(original[0].sortOrder).toBeUndefined();
  expect(sortStyles(moveStyle(moved,'a','b'),'custom').map(x=>x.id)).toEqual(['a','b','c']);
  expect(moveStyle(original,'missing','b')).toBe(original);
 });
 it('imports one raw prompt per line without changing weights and supports named tabs',()=>{
  expect(parseStyleLines('\uFEFF1.2::artist:a::, artist:b\r\n\nDream\t0.8::artist:c::')).toEqual([{name:'1.2::artist:a::, artist:b',prompt:'1.2::artist:a::, artist:b'},{name:'Dream',prompt:'0.8::artist:c::'}]);
 });
 it('matches natural filename order, exact names, no-match and rejects ambiguous files',()=>{
  const names=['10.png','2.png','Dream.jpg'];const rows=[{name:'dream'},{name:'missing'}];
  expect(matchStyleImages(names,rows,'name')).toEqual([['Dream.jpg'],[]]);
  expect(matchStyleImages(names,rows,'order')).toEqual([['2.png'],['10.png']]);
  expect(matchStyleImages(names,rows,'none')).toEqual([[],[]]);
  expect(()=>matchStyleImages(['a.png','a.png'],rows,'order')).toThrow('Duplicate');
 });
 it('all five UI languages cover every sort choice',async()=>{const {default:labels}=await import('../shared/style-library-ui.json');for(const t of Object.values(labels)){for(const s of STYLE_SORTS)expect(t[s==='name'?'nameSort':s]).toBeTruthy();}});
});

describe('style cover selection',()=>{
 const previewImages=['first','second','third'].map(id=>({id,name:id,filePath:id,fileUrl:id,createdAt:''}));
 it('preserves image order and selected cover through JSON roundtrip',()=>{
  const p={...make('a'),previewImages,coverImageId:'second'};
  expect(styleCoverIndex(JSON.parse(JSON.stringify(p)))).toBe(1);
  expect(previewImages.map(im=>im.id)).toEqual(['first','second','third']);
 });
 it('falls back to first image for legacy, deleted or missing covers and handles empty',()=>{
  expect(styleCoverIndex({previewImages})).toBe(0);
  expect(styleCoverIndex({previewImages,coverImageId:'deleted'})).toBe(0);
  expect(styleCoverIndex({previewImages:[],coverImageId:'second'})).toBe(-1);
  expect(styleCoverIndex({})).toBe(-1);
 });
});

File: mobile/test/style_library_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:novelai_mobile/models/nai_models.dart';
import 'package:novelai_mobile/models/style_library.dart';
import 'package:novelai_mobile/state/app_state.dart';
import 'package:shared_preferences/shared_preferences.dart';

StylePromptPreset p(String id,
        {double rating = 0, int uses = 0, String date = '2026-09-01'}) =>
    StylePromptPreset(
        id: id,
        name: id,
        prompt: 'artist:$id',
        createdAt: date,
        rating: rating,
        usageCount: uses);
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  test('old presets, free decimal rating, persistence', () {
    final old = StylePromptPreset.fromJson(
        {'id': 'a', 'name': 'a', 'prompt': 'artist:a', 'createdAt': ''});
    expect(old.rating, 0);
    expect(old.usageCount, 0);
    expect(old.sortOrder, 9007199254740991);
    final value = StylePromptPreset.fromJson(
        {...old.toJson(), 'rating': 4.26, 'usageCount': 3.9, 'sortOrder': 2});
    expect(value.rating, 4.26);
    expect(value.usageCount, 3);
    expect(StylePromptPreset.fromJson(value.toJson()).sortOrder, 2);
    expect(
        StylePromptPreset.fromJson({...old.toJson(), 'rating': 99}).rating, 5);
  });
  test('cover persistence, legacy fallback, removed cover and empty list', () {
    final value = p('a');
    value.previewImages = List.generate(3, (i) => StylePromptPreviewImage(id: 'im$i', name: '$i', filePath: '/$i', createdAt: ''));
    expect(value.coverIndex, 0);
    value.coverImageId = 'im2';
    final restored = StylePromptPreset.fromJson(value.toJson());
    expect(restored.coverImageId, 'im2');
    expect(restored.coverIndex, 2);
    expect(restored.previewImages.map((im) => im.id).join(','), 'im0,im1,im2');
    restored.previewImages.removeAt(2);
    expect(restored.coverIndex, 0);
    expect(StylePromptPreset.fromJson(restored.toJson()).coverImageId, 'im0');
    restored.previewImages.clear();
    expect(restored.coverIndex, -1);
    expect(restored.toJson()['coverImageId'], isNull);
  });
  test('all sorting modes and stable ties', () {
    final items = [
      p('b', rating: 1, uses: 2),
      p('a', rating: 5, date: '2026-09-03'),
      p('c', rating: 1, uses: 3, date: '2026-09-02')
    ];
    final expected = {
      'default': 'bac',
      'custom': 'bac',
      'rating-desc': 'abc',
      'rating-asc': 'bca',
      'uses-desc': 'cba',
      'uses-asc': 'abc',
      'created-desc': 'acb',
      'created-asc': 'bca',
      'name': 'abc'
    };
    for (final mode in styleSorts) {
      expect(sortStyles(items, mode).map((p) => p.id).join(), expected[mode],
          reason: mode);
    }
    expect(items.map((p) => p.id).join(), 'bac');
  });
  test('custom order preserves default sequence', () {
    final items = ['a', 'b', 'c'].map(p).toList();
    moveStyle(items, 'a', 'c');
    expect(sortStyles(items, 'custom').map((p) => p.id).join(), 'bca');
    expect(items.map((p) => p.id).join(), 'abc');
    moveStyle(items, 'a', 'b');
    expect(sortStyles(items, 'custom').map((p) => p.id).join(), 'abc');
  });
  test('batch prompt and image matching', () {
    final rows =
        parseStyleLines('\uFEFF1.2::artist:a::\r\n\nDream\t0.8::artist:b::');
    expect(rows.length, 2);
    expect(rows[1], (name: 'Dream', prompt: '0.8::artist:b::'));
    expect(
        matchStyleImages(
            ['10.png', '2.png', 'Dream.jpg'], ['dream', 'missing'], 'name'),
        [
          ['Dream.jpg'],
          []
        ]);
    expect(matchStyleImages(['10.png', '2.png'], ['a', 'b'], 'order'), [
      ['2.png'],
      ['10.png']
    ]);
    expect(() => matchStyleImages(['a.png', 'a.png'], ['a'], 'name'),
        throwsFormatException);
  });
  test('queued updates preserve both rating and usage, persist after reload',
      () async {
    SharedPreferences.setMockInitialValues({});
    final app = AppState();
    try {
      app.settings.stylePromptPresets = [p('a')];
      await Future.wait([
        app.updateStyleLibrary((s) => s.stylePromptPresets[0].rating = 4.5),
        app.updateStyleLibrary((s) => s.stylePromptPresets[0].usageCount++)
      ]);
      expect(app.settings.stylePromptPresets.single.rating, 4.5);
      expect(app.settings.stylePromptPresets.single.usageCount, 1);
      final saved = await app.storage.getSettings();
      expect(saved.stylePromptPresets.single.rating, 4.5);
      expect(saved.stylePromptPresets.single.usageCount, 1);
    } finally {
      app.dispose();
    }
  });
}
